lib/spack/spack/spec.py

```diff
@@ -1670,7 +1670,14 @@ def to_node_dict(self, hash=ht.dag_hash):
                 d['patches'] = variant._patches_in_order_of_appearance
 
         if hash.package_hash:
-            package_hash = self.package_hash()
+            # Guard against computing package_hash if this spec was read
+            # from disk.  If the serialized version did not have these
+            # hashes, it shouldn't hurt anything to just set them to the
+            # same value as the dag_hash.
+            if self._hashes_final and not getattr(self, hash.attr, None):
+                package_hash = getattr(self, ht.dag_hash.attr, None)
+            else:
+                package_hash = self.package_hash()
```

Comment:
It's weird to me that we're assigning `dag_hash` to `package_hash`.  When we discussed this, I said that we should promote `dag_hash` to `build_hash` to `full_hash`, mainly b/c we sometimes need to index things by those and the weaker hash is still the best identifier we have for the spec.  `package_hash`, though, is a content has of the `package.py` files.  We still don't actually know it -- if we can get away with just leaving it `None` I'd prefer to. 

Do we need to set this if we can instead set `full_hash` and `build_hash`?

Reply:
It's the computation of the `package_hash` that's causing the problems right now, let me see if I can just leave it `None`.

Reply:
I think the only place we'd ask for it to be computed is when getting the full hash if the full hash is not set.

Reply:
Actually, now I'm wondering: Let's say I have an old `spec.yaml` (now they're `spec.json`, but this one's old) for pkg `a` which depends on some dag hash of pkg `b`.  Recently `b` was rebuilt with a new full hash, but because the dag hash of `b` didn't change, I need to splice the new `b` into `a`.  Doesn't it mean I need to recompute the full hash of `a` now?  Because it's the serialized spec dict that's getting hashed, and if the full hash of a dep of `a` changes, so will the full hash of `a` itself?

Reply:
I guess that's where promoting the `dag_hash` is intended to help.